### PR TITLE
Moving the registry to the base backend, and defining its public API

### PR DESCRIPTION
**Pull request by datapythonista**
Working on consolidating the base backends in a meaningful way, and in a way that it's clear what is its public API that other backends can use.

In this PR I move the registry to the base backend (from `base_sql`, which I expect to remove soon). I split the registry in different files, so it's easier to navigate its code. And I make specific what is public, and what it's not.

No functionality has been modified here, just moved code.

In a follow up I'll do the same with the `base_sql` compiler code, and then work on integrating the compiler code in `base_sqlalchemy` into `base`.

### CODE REVIEW

File: ibis/backends/base/__init__.py [hunk 1]
```diff
@@ -2,6 +2,29 @@
 
 from ibis.common.exceptions import TranslationError
 
+from .registry import binary_infix_ops, fixed_arity, operation_registry, unary
```
Comment (jreback) [inline ibis/backends/base/__init__.py:5]: why are you directly importing these here as opposed to from the backends import what is needed?

the reason i ask is that this is common to all backends whereas the registry is for sql only

**Review by jreback — CHANGES_REQUESTED** [review]

File: ibis/backends/base/__init__.py (hunk 1 above)
Comment (datapythonista) [reply ibis/backends/base/__init__.py:5]: Good point. I was focusing on better structuring the code, and defining what was public, and just did this for simplicity.

What do you think about having a `ibis.backends.base.sql` module where these are made available instead?

Comment (jreback) [reply ibis/backends/base/__init__.py:5]: sure that would be great.

**Comment by jreback** [comment]
https://github.com/ibis-project/ibis/pull/2718/checks?check_run_id=2247567616 :->

this is why precommits are helpful on the user side

**Comment by datapythonista** [comment]
> https://github.com/ibis-project/ibis/pull/2718/checks?check_run_id=2247567616 :->
> 
> this is why precommits are helpful on the user side

hahaha, you got me here. ;)

But still, I don't like magic in software. For me the problem is using black and isort, not the lack of precommit. ;) Will have a look anyway after the release, but don't count I'll change my mind.

Anyway, this is green now.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @datapythonista nice reorg, keep em coming!

**Pull request merged**